solutions/AC Matching.mdx

```diff
@@ -0,0 +1,59 @@
+---
+id: ac-matching
+source: AtCoder DP Contest
+title: Matching
+author: Neo Wang
+---
+
+## Explanation
+
+If we define $\texttt{dp}[S]$ to be the amount of matchings for females in set
+$S$ to the first $|S|$ males, this problem boils down to the following
+formulation:
+
+$$
+\texttt{dp}[S] = \sum \texttt{dp}[S\backslash x]\: \text{if}\:  \texttt{compatible}[|S|][x]
+$$
+
+In English, this is the equivalent of
+
+> The amount of matchings in a subset $S$ to include a certain female $x$ is
+> equivalent to the sum of all the matchings without female $x$ such that the
+> addition of female $x$ is compatible with the $|S|$-th male.
+
+Our base case is the empty set, which has a value of $1$. Although the empty set
```

Comment:
> Although the empty set itself contains zero matchings

why wouldn't the empty set be considered to generate a single (empty) matching?

Reply:
isn't dp[s] the amount of matchings for females in set|s| to the first s males? If so then wouldn't the empty set contain zero matchings as there are no females. Also doesn't "generating the base cases" indicate that the empty set is used to generate a single matching.

Reply:
aka "it generates a matching, but by itself does not contain any match" which is valid because N>=1"

Reply:
> isn't dp[s] the amount of matchings for females in set|s| to the first s males?

yes

> If so then wouldn't the empty set contain zero matchings as there are no females

Er, I would consider the empty set as a single matching involving zero pairs.

Reply:
oh that's pretty neat.